pip/configuration.py

```diff
@@ -1,126 +1,317 @@
 """Configuration management setup
+
+Some terminology:
+- name
+  As written in config files.
+- value
+  Value associated with a name
+- key
+  Name combined with it's section (section.name)
+- variant
+  A single word describing where the configuration key-value pair came from
 """
 
+import logging
 import os
-import re
-import sys
 
+from pip._vendor.six import next
 from pip._vendor.six.moves import configparser
 
+from pip.exceptions import ConfigurationError
 from pip.locations import (
-    config_basename, legacy_config_file, running_under_virtualenv,
-    site_config_files
+    legacy_config_file, new_config_file, running_under_virtualenv,
+    site_config_files, venv_config_file
 )
-from pip.utils import appdirs
+from pip.utils import ensure_dir, enum
+
+_need_file_err_msg = "Needed a specific file to be modifying."
+
 
-_environ_prefix_re = re.compile(r"^PIP_", re.I)
+logger = logging.getLogger(__name__)
+
+
+# NOTE: Maybe use the optionx attribute to normalize keynames.
+def _normalize_name(name):
+    """Make a name consistent regardless of source (environment or file)
+    """
+    name = name.lower().replace('_', '-')
+    if name.startswith('--'):
+        name = name[2:]  # only prefer long opts
+    return name
+
+
+def _disassemble_key(name):
+    return name.split(".", 1)
+
+
+# The kinds of configurations there are.
+kinds = enum(
+    USER="user", GLOBAL="global", VENV="venv", ENV="env", ENV_VAR="env-var"
+)
 
 
 class Configuration(object):
-    """Handles the loading of configuration files and providing an interface to
-    accessing data within them.
+    """Handles management of configuration.
+
+    Provides an interface to accessing and managing configuration files.
+
+    This class converts provides an API that takes "section.key-name" style
+    keys and stores the value associated with it as "key-name" under the
+    section "section".
+
+    This allows for a clean interface wherein the both the section and the
+    key-name are preserved in an easy to manage form in the configuration files
+    and the data stored is also nice.
     """
 
-    def __init__(self, isolated):
-        self._configparser = configparser.RawConfigParser()
-        self._config = {}
+    def __init__(self, isolated, load_only=None):
+        super(Configuration, self).__init__()
+
+        _valid_load_only = [kinds.USER, kinds.GLOBAL, kinds.VENV, None]
+        if load_only not in _valid_load_only:
+            raise ConfigurationError(
+                "Got invalid value for load_only - should be one of {}".format(
+                    ", ".join(map(repr, _valid_load_only[:-1]))
+                )
+            )
         self.isolated = isolated
+        self.load_only = load_only
 
-    def load(self, section):
-        """Loads configuration
+        # The order here determines the override order.
+        self._override_order = [
+            kinds.GLOBAL, kinds.USER, kinds.VENV, kinds.ENV, kinds.ENV_VAR
```

Comment:
A comment explaining the difference between kinds.ENV (the config from the file from PIP_CONFIG_FILE) and kinds.ENV_VAR (the config from environment vars) would have helped :o